reverted:
--- b/drivers/mmc/host/sdhci-omap.c
+++ a/drivers/mmc/host/sdhci-omap.c
@@ -1343,9 +1343,6 @@
 	/* R1B responses is required to properly manage HW busy detection. */
 	mmc->caps |= MMC_CAP_NEED_RSP_BUSY;
 
-	/* Allow card power off and runtime PM for eMMC/SD card devices */
-	mmc->caps |= MMC_CAP_POWER_OFF_CARD | MMC_CAP_AGGRESSIVE_PM;
-
 	ret = sdhci_setup_host(host);
 	if (ret)
 		goto err_rpm_put;
