--- a/drivers/rtc/Kconfig
+++ b/drivers/rtc/Kconfig
@@ -158,6 +158,10 @@ config RTC_DRV_TEST
 	  This driver can also be built as a module. If so, the module
 	  will be called rtc-test.
 
+config RTC_DRV_VRTC
+	tristate "Virtual RTC Timer driver"
+	help
+
 comment "I2C RTC drivers"
 
 if I2C
--- a/drivers/rtc/Makefile
+++ b/drivers/rtc/Makefile
@@ -179,6 +179,7 @@ obj-$(CONFIG_RTC_DRV_TPS6586X)	+= rtc-tp
 obj-$(CONFIG_RTC_DRV_TPS65910)	+= rtc-tps65910.o
 obj-$(CONFIG_RTC_DRV_TWL4030)	+= rtc-twl.o
 obj-$(CONFIG_RTC_DRV_V3020)	+= rtc-v3020.o
+obj-$(CONFIG_RTC_DRV_VRTC)      += rtc-vrtc.o
 obj-$(CONFIG_RTC_DRV_VT8500)	+= rtc-vt8500.o
 obj-$(CONFIG_RTC_DRV_WILCO_EC)	+= rtc-wilco-ec.o
 obj-$(CONFIG_RTC_DRV_WM831X)	+= rtc-wm831x.o
