--- a/drivers/net/ethernet/ti/cpsw_new.c
+++ b/drivers/net/ethernet/ti/cpsw_new.c
@@ -720,6 +720,7 @@ static void cpsw_slave_open(struct cpsw_
 {
 	struct cpsw_common *cpsw = priv->cpsw;
 	struct phy_device *phy;
+	__ETHTOOL_DECLARE_LINK_MODE_MASK(mask) = { 0, };
 
 	cpsw_sl_reset(slave->mac_sl, 100);
 	cpsw_sl_ctl_reset(slave->mac_sl);
@@ -776,6 +777,13 @@ static void cpsw_slave_open(struct cpsw_
 	}
 	slave->phy = phy;
 
+	/* disable adv 1G by default */
+	dev_info(priv->dev, "slave %d: disable 1G advertising by default", 
+		slave->slave_num);
+	linkmode_set_bit(ETHTOOL_LINK_MODE_1000baseT_Full_BIT, mask);
+	linkmode_set_bit(ETHTOOL_LINK_MODE_1000baseT_Half_BIT, mask);
+	linkmode_andnot(slave->phy->advertising, slave->phy->advertising, mask);
+
 	phy_attached_info(slave->phy);
 
 	phy_start(slave->phy);
